libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt

```diff
@@ -130,695 +103,378 @@ internal object MapboxNavigationTelemetry : MapboxNavigationTelemetryInterface {
-        retVal.complete(true)
-        return retVal
     }
 
-    /**
-     * This method clears the state data for the Telemetry object in response to onSessionEnd
-     */
-    private fun handleSessionStop() {
-        dynamicValues.reset()
-        callbackDispatcher.clearOriginalRoute()
+    private fun registerListeners(mapboxNavigation: MapboxNavigation) {
+        mapboxNavigation.run {
+            registerLocationObserver(locationsCollector)
+            registerRouteProgressObserver(this@MapboxNavigationTelemetry)
+            registerRoutesObserver(this@MapboxNavigationTelemetry)
+            registerOffRouteObserver(this@MapboxNavigationTelemetry)
+            registerNavigationSessionObserver(this@MapboxNavigationTelemetry)
+        }
     }
 
-    /**
-     * This method starts a session. If a session is active it will terminate it, causing an stop/cancel event to be sent to the servers.
-     * Every session start is guaranteed to have a session end.
-     */
-    private fun handleSessionStart() {
-        telemetryThreadControl.scope.launch {
-            Log.d(TAG, "Waiting in handleSessionStart")
-            dynamicValues.originalRoute.set(callbackDispatcher.getOriginalRouteAsync().await())
-            dynamicValues.originalRoute.get()?.let { directionsRoute ->
-                Log.d(TAG, "The wait is over")
-                sessionStartHelper(directionsRoute)
-            }
+    fun unregisterListeners(mapboxNavigation: MapboxNavigation) {
+        mapboxNavigation.run {
+            unregisterLocationObserver(locationsCollector)
+            unregisterRouteProgressObserver(this@MapboxNavigationTelemetry)
+            unregisterRoutesObserver(this@MapboxNavigationTelemetry)
+            unregisterOffRouteObserver(this@MapboxNavigationTelemetry)
+            unregisterNavigationSessionObserver(this@MapboxNavigationTelemetry)
         }
+        MapboxMetricsReporter.disable()
     }
 
-    /**
-     * This method is used by a lambda. Since the Telemetry class is a singleton, U.I. elements may call postTurnstileEvent() before the singleton is initialized.
-     * A lambda guards against this possibility
-     */
-    private fun postTurnstileEvent() {
-        // AppUserTurnstile is implemented in mapbox-telemetry-sdk
-        val sdkId = generateSdkIdentifier()
-        dynamicValues.sdkId = generateSdkIdentifier()
-        val appUserTurnstileEvent =
-            AppUserTurnstile(sdkId, BuildConfig.MAPBOX_NAVIGATION_VERSION_NAME).also {
-                it.setSkuId(
-                    MapboxNavigationAccounts.getInstance(
-                        context
-                    ).obtainSkuId()
-                )
+    private fun NavigationEvent.populate(): NavigationEvent {
+        log("populateNavigationEvent")
+
+        return this.apply {
+            sdkIdentifier = this@MapboxNavigationTelemetry.sdkIdentifier
+
+            routeProgress!!.let { routeProgress ->
```

Comment:
Should we check for `routeProgress` and `originalRoute` first to populate all the data? Wondering if we could run into a situation in which `routeProgress` and `originalRoute` are `null` but the event is sent with missing / partial data 🤔

Reply:
no need to check it.
`NavigationEvent.populate()` can be called only if the session is started.
And session can be started when `routeProgress` and `originalRoute` are not null.

> Wondering if we could run into a situation in which routeProgress and originalRoute are null but the event is sent with missing / partial data 🤔

if any of them is null we will have a crash, no event will be sent.